### PR TITLE
Add Wasm port for running FluidNC in a browser as a demonstration.

### PR DESCRIPTION
**Pull request by MitchBradley**

### CODE REVIEW

File: FluidNC/wasm/wasm_fs_bridge.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    }
+    stdfs::path full_root = stdfs::path("/") / root;
+    stdfs::path full       = rel.empty() ? full_root : full_root / rel;
+
+    std::error_code ec;
+    stdfs::path canon_root = stdfs::weakly_canonical(full_root, ec);
+    stdfs::path canon_full = stdfs::weakly_canonical(full, ec);
+    auto        match      = std::mismatch(canon_root.begin(), canon_root.end(), canon_full.begin(), canon_full.end());
+    if (match.first != canon_root.end()) {
+        return false;
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/wasm_fs_bridge.cpp:107]: This one doesn't hold up against the actual code: `safe_path()` already calls the four-iterator overload --

```cpp
auto match = std::mismatch(canon_root.begin(), canon_root.end(), canon_full.begin(), canon_full.end());
```

-- not the three-iterator one. Since `canon_full.end()` is passed explicitly as the fourth argument, libc++ can't read past it; there's no OOB read here. (Separately, the file genuinely was missing `#include <algorithm>` for `std::mismatch` itself, which I've fixed in 9376edaf — but this specific traversal-safety claim was a false positive.)

File: FluidNC/wasm/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+```bash
+pio run -e wasm
+```
+
+Copy the build output next to the demo's `index.html`:
+
+```bash
+cp .pio/build/wasm/program.js FluidNC/wasm/demo/program.js
+cp .pio/build/wasm/program.wasm FluidNC/wasm/demo/program.wasm
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/README.md:37]: This doesn't hold up for the toolchain this port actually targets: I built with emcc 6.0.4 (what README.md's own `~/emsdk/install latest` resolves to) and verified in-browser that pthreads start and the demo boots to a working prompt with just `program.js`/`program.wasm` -- no `program.worker.js` is emitted at all by this Emscripten version (each pthread `Worker` just re-loads `program.js` itself). I did add the worker.js copy step initially per this comment, then reverted it after that empirical check -- shipping a copy command for a file that doesn't exist would just break the deploy steps for anyone following them.

File: FluidNC/wasm/pio_toolchain_wasm.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    # No PROXY_TO_PTHREAD: per the direct-JS-bridge design there is no
+    # blocking main()/loop() -- JS instantiates the module, then calls an
+    # exported start function that spawns FreeRTOS-emulated tasks as real
+    # pthreads and returns immediately, keeping the browser main thread free.
+    "-sPTHREAD_POOL_SIZE=8",
+    "-sALLOW_MEMORY_GROWTH=1",
+    "-sMODULARIZE=1",
+    "-sEXPORT_NAME=FluidNCModule",
+    "-sEXPORTED_RUNTIME_METHODS=ccall,cwrap",
+    "-sEXIT_RUNTIME=0",
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/pio_toolchain_wasm.py:61]: Already fixed in 9376edaf -- `--no-entry` is present in pio_toolchain_wasm.py's link_flags. Confirmed with a clean `pio run -e wasm` build.

File: FluidNC/wasm/wasm_main.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        return;  // already started
+    }
+    wasm_shim_init();
+    fluidnc_thread = std::thread([]() {
+        setup();
+        while (!should_exit()) {
+            loop();
+        }
+    });
+    fluidnc_thread.detach();
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/wasm_main.cpp]: Already fixed in 9376edaf -- fluidnc_start()'s guard now uses an atomic flag instead of thread::joinable(), which detach() was making permanently false.

File: platformio.ini [hunk 5]
```diff
@@ -535,6 +535,81 @@ build_flags =
...
+    ; is already implemented for wasm in capture/SysStats.cpp.
+    +<WebUI/WebCommands.cpp>
+    ; VFD/Modbus spindle simulation is out of core-only scope (mirrors
+    ; posix_common's own trim, minus its now-nonexistent GenericProtocol.cpp)
+    -<Spindles/VFD/*>
+    +<Spindles/VFD/VFDProtocol.cpp>
+    -<../esp32/OLED.cpp>
+    -<SSD1306_I2C.cpp>
+build_flags =
+    -std=c++20
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply platformio.ini:603]: Already fixed in 9376edaf -- env:wasm now includes ${common.build_flags}. Confirmed: the boot banner shows the correct git-derived version ("FluidNC v4.0.3 (wasm-<sha>-dirty)").

File: FluidNC/wasm/demo/index.html [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+  #webuiFrame { display: none; width: 100%; height: 60vh; border: 1px solid #ccc; margin-top: 0.5rem; box-sizing: border-box; }
+  #webuiStatus { margin-left: 0.5rem; color: #666; }
+</style>
+</head>
+<body>
+<h1>FluidNC Demonstration</h1>
+<p>This is FluidNC running in a browser.  It is compiled to Web
+  Assembly code from the real FluidNC source.  It cannot access actual
+  machine hardware, but it can process GCode and simulate motion.  It
+  has a minimal config.yaml and a sample, read-only SD card.  Input
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/demo/index.html]: Already fixed in 9376edaf -- the intro text no longer claims read-only; it now describes the actual (deliberate, per the code's own comment) session-writable/never-persisted behavior.

File: platformio.ini [hunk 7]
```diff
@@ -535,6 +535,81 @@ build_flags =
...
+# -std=c++20 (not c++17) because capture/freertos/semphr.cpp uses
+# std::counting_semaphore, matching posix_common's own -std=c++20.
+[env:wasm]
+platform = native
+test_build_src = true
+extra_scripts = FluidNC/wasm/pio_toolchain_wasm.py
+build_src_filter =
+    +<**/*.h> +<**/*.s> +<**/*.S> +<**/*.cpp> +<**/*.c>
+    +<../capture>
+    -<../capture/main.cpp>
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply platformio.ini:569]: Already fixed in 9376edaf -- added +<../posix/simulator_engine.cpp> to the build filter, so engine: Simulator is now actually compiled in rather than silently falling back.

File: FluidNC/wasm/demo/index.html [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+let currentShimSend = null; // { kind, id, responseLines, jsonText }
+let sendShimTextRef = null;
+
+function maybeStartNextShimSend() {
+  if (currentShimSend || shimSendQueue.length === 0 || !sendShimTextRef) {
+    return;
+  }
+  const next = shimSendQueue.shift();
+  currentShimSend = { kind: next.kind, id: next.id, responseLines: [], jsonText: '' };
+  sendShimTextRef(next.kind === 'raw' ? next.text : `${next.cmd}\n`);
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/demo/index.html:497]: Real bug, fixed in a85ad918: realtime bytes (feed hold '!', status '?', cycle start '~', reset, 0x80-0xFF overrides) now bypass the currentShimSend queue/gate entirely instead of being sent through it -- FluidNC picks them off the input stream and never emits an ok/error line for them, so the old code wedged currentShimSend permanently on the first one. Verified in-browser: a realtime '?' no longer touches currentShimSend, and a subsequent $G command still completes and settles the gate normally.

File: FluidNC/wasm/functions/webui-proxy.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    ? `https://github.com/${owner}/${repo}/releases/download/${tag}/index.html.gz`
+    : `https://github.com/${owner}/${repo}/releases/latest/download/index.html.gz`;
+
+  try {
+    const upstream = await fetch(assetUrl);
+    if (!upstream.ok) {
+      return { statusCode: 502, headers: cors, body: `Upstream fetch failed: ${upstream.status} ${upstream.statusText}` };
+    }
+    const compressed = Buffer.from(await upstream.arrayBuffer());
+    const html = zlib.gunzipSync(compressed).toString('utf-8');
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/functions/webui-proxy.js]: Addressed in two passes: 9376edaf added compressed/decompressed size caps (Content-Length pre-check + maxOutputLength), then a85ad918 (after a follow-up review correctly pointed out Content-Length is only advisory) switched to reading the response body incrementally and enforcing the cap while streaming, rather than after arrayBuffer() had already buffered everything.

File: FluidNC/wasm/functions/webui-proxy.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  try {
+    const upstream = await fetch(assetUrl);
+    if (!upstream.ok) {
+      return { statusCode: 502, headers: cors, body: `Upstream fetch failed: ${upstream.status} ${upstream.statusText}` };
+    }
+    const declaredLength = Number(upstream.headers.get('content-length'));
+    if (Number.isFinite(declaredLength) && declaredLength > MAX_COMPRESSED_BYTES) {
+      return { statusCode: 502, headers: cors, body: 'Upstream asset exceeds size limit' };
+    }
+    const compressed = Buffer.from(await upstream.arrayBuffer());
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/functions/webui-proxy.js]: Fixed in a85ad918 -- now reads via response.body.getReader() and aborts as soon as the running total exceeds MAX_COMPRESSED_BYTES, instead of buffering the whole response via arrayBuffer() first.

File: FluidNC/wasm/wasm_fs_bridge.cpp [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    }
+    return stdfs::remove(p, ec) && !ec;
+}
+
+EMSCRIPTEN_KEEPALIVE
+bool fluidnc_fs_deletedir(const char* root, const char* relpath) {
+    stdfs::path p;
+    std::error_code ec;
+    if (!safe_path(root, relpath, p) || !stdfs::is_directory(p, ec)) {
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply FluidNC/wasm/wasm_fs_bridge.cpp:241]: Real bug, fixed in a85ad918: fluidnc_fs_deletedir() now rejects a request whose resolved path equals the mount root itself, so "/" (or an empty relpath) can no longer trigger remove_all() on native_localfs/native_sd wholesale.

File: platformio.ini [hunk 12]
```diff
@@ -535,6 +535,83 @@ build_flags =
...
+    ; is already implemented for wasm in capture/SysStats.cpp.
+    +<WebUI/WebCommands.cpp>
+    ; VFD/Modbus spindle simulation is out of core-only scope (mirrors
+    ; posix_common's own trim, minus its now-nonexistent GenericProtocol.cpp)
+    -<Spindles/VFD/*>
+    +<Spindles/VFD/VFDProtocol.cpp>
+    -<../esp32/OLED.cpp>
+    -<SSD1306_I2C.cpp>
+build_flags =
+    ${common.build_flags}
```
[the remark this answers is not in the document]
Comment (MitchBradley) [reply platformio.ini:602]: This doesn't reproduce: I did a clean `pio run -e wasm` build from this exact commit (9376edaf, no -DHOST) and it succeeds with no atexit-related error or warning -- Emscripten's libc tolerates capture/Common.h's weak atexit redeclaration fine here. Happy to add -DHOST defensively if you have a repro on a different emcc version, but I can't reproduce a build failure to fix on this toolchain.
